src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,244 @@
+        {
+        }
+
+        /// <summary>
+        /// The MSBuild properties used when interpreting project files.
+        /// These are the same properties that are passed to msbuild via the /property:&lt;n&gt;=&lt;v&gt; command line argument.
+        /// </summary>
+        public ImmutableDictionary<string, string> Properties => _properties;
+
+        /// <summary>
+        /// Determines if metadata from existing output assemblies is loaded instead of opening referenced projects.
+        /// If the referenced project is already opened, the metadata will not be loaded.
+        /// If the metadata assembly cannot be found the referenced project will be opened instead.
+        /// </summary>
+        public bool LoadMetadataForReferencedProjects { get; set; } = false;
+
+        /// <summary>
+        /// Determines if unrecognized projects are skipped when solutions or projects are opened.
+        /// 
+        /// A project is unrecognized if it either has 
+        ///   a) an invalid file path, 
+        ///   b) a non-existent project file,
+        ///   c) has an unrecognized file extension or 
+        ///   d) a file extension associated with an unsupported language.
+        /// 
+        /// If unrecognized projects cannot be skipped a corresponding exception is thrown.
+        /// </summary>
+        public bool SkipUnrecognizedProjects { get; set; } = true;
+
+        /// <summary>
+        /// Associates a project file extension with a language name.
+        /// </summary>
+        public void AssociateFileExtensionWithLanguage(string projectFileExtension, string language)
+        {
+            if (language == null)
+            {
+                throw new ArgumentNullException(nameof(language));
+            }
+
+            if (projectFileExtension == null)
+            {
+                throw new ArgumentNullException(nameof(projectFileExtension));
+            }
+
+            _projectFileLoaderRegistry.AssociateFileExtensionWithLanguage(projectFileExtension, language);
+        }
+
+        private void SetSolutionProperties(string solutionFilePath)
+        {
+            const string SolutionDirProperty = "SolutionDir";
+
+            // When MSBuild is building an individual project, it doesn't define $(SolutionDir).
+            // However when building an .sln file, or when working inside Visual Studio,
+            // $(SolutionDir) is defined to be the directory where the .sln file is located.
+            // Some projects out there rely on $(SolutionDir) being set (although the best practice is to
+            // use MSBuildProjectDirectory which is always defined).
+            if (!string.IsNullOrEmpty(solutionFilePath))
+            {
+                var solutionDirectory = Path.GetDirectoryName(solutionFilePath);
+                if (!solutionDirectory.EndsWith(@"\", StringComparison.Ordinal))
+                {
+                    solutionDirectory += @"\";
```

Comment:
>@"\"; [](start = 41, length = 5)

Rather than hard coding `\` should we use `Path.DirectorySeparatorChar`?

Reply:
Yes, we should do that now to prepare for eventually running cross platform. Good catch!

Reply:
Also, the check to see if `solutionDirectory` ends with `'\'` is unnecessary since `Path.GetDirectoryName(...)` doesn't return the trailing slash.